### PR TITLE
Feature query annotation

### PR DESCRIPTION
**Pull request by helpotters**
- annotate queries that pass through `Readyset.route`
- tagged with `/*routed_to_readyset?:true*/`

The feature below is made redundant by having `query_logs_enabled` set to `true`.
- ~toggle this feature on using `config.query_annotations = true`~

### CODE REVIEW

**Comment by helpotters** [comment]
I'm sharing the failing code. I'm having a major difficulty with the testing environment setup with enabling `QueryLogs`. By default, it should at least add `/*application:MyApp*/` to every query in `test.log`.

File: lib/readyset/query_annotator.rb
Comment (ethan-readyset) [inline lib/readyset/query_annotator.rb:10]: Hmm `ActiveSupport::ExecutionContext` is an [undocumented class](https://github.com/rails/rails/blob/6b93fff8af32ef5e91f4ec3cfffb081d0553faf0/activesupport/lib/active_support/execution_context.rb#L4) in Rails, so we probably shouldn't use it here -- it could be changed/removed with minor Rails version updates.

I think we can integrate the way that the `ActiveRecord::QueryLogs` documentation suggests: api.rubyonrails.org/classes/ActiveSupport/QueryLogs.html
> New comment tags can be defined by adding them in a Hash to the tags Array. Tags can have dynamic content by setting a Proc or lambda value in the Hash, and can reference any value stored by Rails in the context object. [ActiveSupport::CurrentAttributes](https://api.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html) can be used to store application values. Tags with nil values are omitted from the query comment.

Basically, we could do something like this to pass info about whether we're routing to ReadySet:
```ruby
Rails.configuration.active_record.query_log_tags << {
  routed_to_readyset?: ->(context) do
    Current.routing_to_readyset?
  end,
}
```
Where `Current` is defined like so:
```ruby
module Readyset
  class Current < ActiveSupport::CurrentAttributes
    attr_writer :routing_to_readyset

    def routing_to_readyset?
      !!@routing_to_readyset
    end
  end
end
```
Then we'd just need to update the `routing_to_readyset` attribute whenever `Readyset.route` is called:
```ruby
  def self.route(prevent_writes: true, &block)
    Current.routing_to_readyset = true
    ...
  ensure
    Current.routing_to_readyset = false
  end
```

Comment (helpotters) [reply lib/readyset/query_annotator.rb:10]: Hmm. I'm still figuring out how to check that the SQL comment was added. 

So far, it's been a struggle trying to get it to output the SQL query log with comments enabled. It might be a limitation with Combustion.

Comment (ethan-readyset) [reply lib/readyset/query_annotator.rb:10]: I'd say we should probably switch to integrating via `ActiveRecord::QueryLogs` before spending too much time debugging the specs. It's possible there's something about using `ExecutionContext` that's causing the spec issue

Comment (helpotters) [reply lib/readyset/query_annotator.rb:10]: Yesterday it wasn't printing `/*application:Combustion*/` in `test.log` queries, but it is today! 

image


(╯°□°)╯︵ ┻━┻

File: spec/ready_set_spec.rb
Comment (ethan-readyset) [inline spec/ready_set_spec.rb:264]: I think we'll want to package up this logging config neatly so users of the gem can enable it without much hassle. What are your thoughts on doing something like this in a railtie?
```ruby
initializer 'readyset.query_annotator' do |app|
  ActiveSupport.on_load(:after_intialization) do
    if Readyset.config.enable_query_annotations?
        if Rails.configuration.active_record.query_log_tags_enabled
          Rails.configuration.active_record.query_log_tags ||= []
          Rails.configuration.active_record.query_log_tags << {
            routed_to_readyset?: ->(context) do
              Readyset::QueryAnnotator.routing_to_readyset?
            end,
          }
      else
        Rails.configuration.active_record.query_log_tags_enabled = true
        Rails.configuration.active_record.query_log_tags = [{
          routed_to_readyset?: ->(context) do
              Readyset::QueryAnnotator.routing_to_readyset?
            end,
          }]
      end
    end
  end
end
```
Basically, if query log tagging is already enabled, we just add our special attribute to the list of tags, and if it's not already enabled, we set the list of tags to consist of _only_ our tag. What do you think?

I want to make the gem as easy to use as possible with as little configuration as possible 😄

File: spec/configuration_spec.rb
Comment (ethan-readyset) [inline spec/configuration_spec.rb:1]: Were the changes in this file added intentionally?

File: lib/readyset.rb
Comment (ethan-readyset) [inline lib/readyset.rb:110]: Can we add some test coverage to check that `Readyset.route` sets `routing_to_readyset` to true at the start of the method and then always sets it to false at the end of the method (even in the case of an error)?

**Review by ethan-readyset — CHANGES_REQUESTED** [review]
Looks great! 🙌  Just a few questions/suggestions

File: spec/ready_set_spec.rb
Comment (helpotters) [reply spec/ready_set_spec.rb:264]: I was on the fence about this due to changing things on their Rails app. But appending the Readyset context if they have `query_log_tags_enabled` is an excellent idea. 

I'll implement this.

Comment (helpotters) [reply spec/ready_set_spec.rb:264]: I was considering this, but decided against it to avoid changing their application state "invisibly". However, I think your example actually makes it work out.

I think it's a good idea to append our context to `query_log_tags` after checking if they have `query_log_tags_enabled`. However, I don't think we should enable it for them, at most, we should warn them that it is disabled/unavailable.

This is a great quality-of-life suggestion.

Comment (helpotters) [reply spec/ready_set_spec.rb:264]: Currently having difficulties with Combustion due to how the initializer order works. It's difficult to test the negative condition of when query tags aren't available. May refactor the Combustion initialization into a support file or something.

Thoughts welcome.

File: spec/configuration_spec.rb
Comment (helpotters) [reply spec/configuration_spec.rb:1]: No :cry:

File: lib/readyset.rb
Comment (helpotters) [reply lib/readyset.rb:110]: Added!

File: spec/ready_set_spec.rb
Comment (helpotters) [reply spec/ready_set_spec.rb:264]: For now, it's just logging warn as a placeholder. What it should really do is use the `Logger` that I gotta make already lol.

Comment (ethan-readyset) [reply spec/ready_set_spec.rb:264]: @helpotters Yeah, I hear you about not changing the application's config. Maybe it would be more palatable if we were to only enable the query log tags when `Rails.env == 'environment'`?

I'm just worried that people won't enable this feature unless they read the docs, and I think it has the potential to make the gem a lot easier to use.

File: lib/readyset/railtie.rb
Comment (ethan-readyset) [inline lib/readyset/railtie.rb:27]: I like this warn idea! I think we could add a little more context -- maybe something like this:
> Query log tags are currently disabled. The readyset gem uses these tags to display information in the logs about whether a query was routed to ReadySet. It is highly recommended that you enable query log tags by setting `Rails.configuration.active_record.query_log_tags_enabled` to `true` so you can verify that queries are being routed to ReadySet as expected.

File: spec/configuration_spec.rb
Comment (ethan-readyset) [inline spec/configuration_spec.rb:10]: I think this change will go away once we rebase this branch on top of main now that #59 has been merged

File: spec/railtie_spec.rb
Comment (ethan-readyset) [inline spec/railtie_spec.rb:52]: I think what's happening here is that the combustion application is already loaded by the time we try to stub this, which means the railtie has already run. We might be able to do this if we explicitly call `Combustion.initialize!` in this `before` block rather than requiring calling it in `spec_helper.rb` (though there may be a better way than this)

**Review by ethan-readyset — CHANGES_REQUESTED** [review]
Looks great! Just a few comments

File: lib/readyset/railtie.rb
Comment (helpotters) [reply lib/readyset/railtie.rb:27]: That's an excellent consideration. I'll be sure to add that.

Comment (helpotters) [reply lib/readyset/railtie.rb]: Adjusted it! It's now operating under a `Rails.env.development?` condition.

File: spec/railtie_spec.rb
Comment (helpotters) [reply spec/railtie_spec.rb]: I managed to work around it by using `Rails.env.development?` as the condition instead.

Lot less prone to failure.

**Comment by helpotters** [comment]
For review, I would recommend double-checking this snippet of code: 
```ruby
    initializer 'readyset.query_annotator' do |app|
      setup_query_annotator
    end

    def setup_query_annotator
      config.after_initialize do
        ...
      end
    end
```

I'm a bit unsure about the usage of `config.after_initialize` as opposed to alternatives.

File: lib/readyset.rb
Comment (ethan-readyset) [inline lib/readyset.rb:92]: Arghhh I realized that this approach won't quite work. It's possible that people may invoke `ActiveRecord::Base.connected_to` within a `Readyset.route` block, which would make us think we were routing to ReadySet when we actually weren't. That said, I discovered an even simpler approach. We can add a query log tag like so:
```ruby
Rails.configuration.active_record.query_log_tags << {
  destination: ->(context) do
    ActiveRecord::Base.connection_db_config.name
  end,
end
```
This will include a `destination` tag that reads either `primary` or `readyset` depending on where the query was headed.

**Review by ethan-readyset — CHANGES_REQUESTED** [review]

**Comment by helpotters** [comment]
That was a rough rebase, but now a spec isn't passing anymore :((. Gotta figure it out.

**Comment by ethan-readyset** [comment]
@helpotters Apologies for rough rebase -- I merged a bunch of big changes to main over the past week or two! The failing spec is my fault -- I can spend some time digging into this later today

**Comment by ethan-readyset** [comment]
@helpotters Fixed in #72

**Comment by helpotters** [comment]
:notes: This is ready for another go :notes:

**Comment by ethan-readyset** [comment]
@helpotters Looks great! Ready to merge after a rebase to resolve the merge conflict

**Review by ethan-readyset — APPROVED** [review]

**Pull request merged**
